src/serialization/sb3.js

```diff
@@ -95,16 +97,59 @@ const primitiveOpcodeInfoMap = {
     data_listcontents: [LIST_PRIMITIVE, 'LIST']
 };
 
+const primitiveIdMap = (() => {
+    const map = {};
+    for (const opcode in primitiveOpcodeInfoMap) {
+        const info = primitiveOpcodeInfoMap[opcode];
+        map[info[0]] = opcode;
+    }
+    return map;
+})();
+
+/**
+ * Retrieve information about the extension providing a particular block.
+ * @param {object} block - a runtime block object, possibly with dynamic blockInfo in a mutation.
+ * @param {Runtime} runtime - the VM runtime.
+ * @returns {ExtensionInfo} - metadata for the extension providing the block, or `undefined` if absent / not loaded.
+ */
+const getExtensionInfoForBlock = function (block, runtime) {
+    const categoryId = block.categoryId ||
+        (block.mutation && block.mutation.blockInfo && block.mutation.blockInfo.categoryId);
+    if (categoryId) {
+        return runtime.getExtensionInfo(categoryId);
+    }
+    // undefined
```

Comment:
cruft?

Reply:
Should we log a warning here if the categoryId was not found?

Reply:
This can happen legitimately for now (until 100% extensionification) so I added some more comment here to explain that.